payments/tests/test_reservation_api.py

```diff
@@ -108,10 +154,112 @@ def test_reservation_that_has_order_cannot_be_modified_without_permission(user_a
+
+
+def test_order_post(user_api_client, resource_in_unit, product, product_2, mock_provider):
+    reservation_data = build_reservation_data(resource_in_unit)
+    reservation_data['order'] = build_order_data(product=product, product_2=product_2, quantity_2=5)
+
+    response = user_api_client.post(LIST_URL, reservation_data)
+
+    assert response.status_code == 201
+    mock_provider.initiate_payment.assert_called()
+
+    # check response fields
+    order_create_response_fields = ORDER_FIELDS.copy() | {'payment_url'}
+    order_data = response.data['order']
+    assert set(order_data.keys()) == order_create_response_fields
+    assert order_data['payment_url'].startswith('https://mocked-payment-url.com')
+
+    # check created object
+    new_order = Order.objects.last()
+    assert new_order.reservation == Reservation.objects.last()
+
+    # check order lines
+    order_lines = new_order.order_lines.all()
+    assert order_lines.count() == 2
+    assert order_lines[0].product == product
+    assert order_lines[0].quantity == 1
+    assert order_lines[1].product == product_2
+    assert order_lines[1].quantity == 5
+
+
+def test_order_product_must_match_resource(user_api_client, product, resource_in_unit, resource_in_unit2):
+    product_with_another_resource = ProductFactory(resources=[resource_in_unit2])
+    data = build_reservation_data(resource_in_unit)
+    data['order'] = build_order_data(product=product, product_2=product_with_another_resource)
+
+    response = user_api_client.post(LIST_URL, data)
+
+    assert response.status_code == 400
+    assert 'product' in response.data['order']['order_lines'][1]
+
+
+def test_order_line_products_are_unique(user_api_client, resource_in_unit, product):
+    """Test order validator enforces that order lines cannot contain duplicates of the same product"""
+    reservation_data = build_reservation_data(resource_in_unit)
+    reservation_data['order'] = build_order_data(product, quantity=2, product_2=product, quantity_2=2)
+    response = user_api_client.post(LIST_URL, reservation_data)
+
+    assert response.status_code == 400
+
+
+@pytest.mark.parametrize('quantity, expected_status', (
+    (1, 201),
+    (2, 201),
+    (3, 400),
+))
+def test_order_line_product_quantity_limitation(user_api_client, resource_in_unit, quantity, expected_status):
+    """Test order validator order line quantity is within product max quantity limitation"""
+    reservation_data = build_reservation_data(resource_in_unit)
+    product_with_quantity = ProductFactory(resources=[resource_in_unit], max_quantity=2)
+    order_data = build_order_data(product=product_with_quantity, quantity=quantity)
+    reservation_data['order'] = order_data
+
+    response = user_api_client.post(LIST_URL, reservation_data)
+
+    assert response.status_code == expected_status, response.data
+
+
+@pytest.mark.parametrize('has_product', (True, False))
+def test_order_mandatoriness(user_api_client, resource_in_unit, has_product):
```

Comment:
It seems "mandatoriness" is a real word although I don't think I've seen it used before. :)

Reply:
Yeah, actually I needed to verify it was a real word as well :D